File: stronghold/015_LIA/lia.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func main() {
	validRunCheck()

	fileContents, err := os.ReadFile(os.Args[1])
	if err != nil {
		fmt.Println("Error reading file, ", err)
		os.Exit(1)
	}

	inputVals := Map(
		Map(
			strings.Split(string(fileContents), " "),
			strings.TrimSpace,
		),
		StrToInt,
	)

	fmt.Println(LIA(inputVals[0], inputVals[1]))
}

func validRunCheck() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run lcsm.go <fasta_file>")
		os.Exit(1)
	}
}

func Map[T, V any](ts []T, fn func(T) V) []V {
	result := make([]V, len(ts))
	for i, t := range ts {
		result[i] = fn(t)
	}
	return result
}

func StrToInt(s string) int {
	i, _ := strconv.Atoi(s)
	return i
}

func Factorial(n int) int {
	f := 1
	for i := n; i > 1; i-- {
		f *= i
	}
	return f
}

func Choose(n, k int) int {
	return Factorial(n) / (Factorial(k) * Factorial(n-k))
}

func BinomialProb(n, k int, p float64) float64 {
	return float64(Choose(n, k)) *
		math.Pow(p, float64(k)) *
		math.Pow((1-p), float64(n-k))
}

func LIA(kthGeneration, N int) float64 {
	var totalProb float64
	popSize := int(math.Pow(2, float64(kthGeneration)))
	fmt.Println(popSize)
	for i := N; i <= popSize; i++ {
		newProb := BinomialProb(popSize, i, 0.25)
		totalProb += newProb
		fmt.Printf("%v Choose %v = %v\n", popSize, i, Choose(popSize, i))
	}
	return totalProb
}

File: stronghold/015_LIA/python/lia.py
#!/usr/bin/env python3
# lia.py

import sys

def factorial(n: int) -> int:
    result = 1
    for i in range(2, n+1):
        result *= i
    return result

def choose(n: int, k: int) -> int:
    return (factorial(n) / (factorial(k) * factorial(n-k)))

def binomial_probability(n: int, k: int, p: float) -> float:
    return choose(n, k) * (p ** k) * ((1 - p) ** (n - k))

def lia(kth_generation: int, at_least: int) -> float:
    total_prob = 0
    pop_size = 2 ** kth_generation

    for i in range(at_least, pop_size+1):
        total_prob += binomial_probability(pop_size, i, 0.25)
    
    return total_prob

def main():
    for line in sys.stdin:
        nums = [int(i) for i in line.strip().split(" ")]
        result = lia(nums[0], nums[1])
        print(f"{result:.3f}")

if __name__ == '__main__':
    main()
